phoenix-core-client/src/main/java/org/apache/phoenix/jdbc/FailoverPhoenixConnection.java

```diff
@@ -333,10 +339,31 @@ <T> T wrapActionDuringFailover(SupplierWithSQLException<T> s) throws SQLExceptio
         while (true) {
             try {
                 return s.get();
-            } catch (SQLException e) {
+            } catch (Exception e) {
+                if (isStaleClusterRoleRecordExceptionExistsInThrowable(e)) {
+                    LOG.debug("StaleHAGroupStoreException found refreshing HAGroup");
+                    //If the exception is due to stale ClusterRoleRecord version, try
+                    //refreshing the ClusterRoleRecord and state transitions if required
+
+                    if (!context.getHAGroup().refreshClusterRoleRecord()) {
+                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.STALE_CRR_RETHROW_AFTER_REFRESH_FAILED)
+                                .setMessage(String.format("Error while running operation Stale ClusterRoleRecord for HAGroup %s" +
+                                                " found with version %s, refreshing HAGroup failed :- %s",
+                                        context.getHAGroup(), context.getHAGroup().getRoleRecord().getVersion(), e)).build().buildException();
+                    }
+
+                    throw new SQLExceptionInfo.Builder(SQLExceptionCode.STALE_CRR_RETHROW_AFTER_REFRESH)
```

Comment:
We come back to the earlier question I had about this implied by its naming. What is an "operation Stale ClusterRoleRecord"? That does not clearly express what is happening here, nor the name of the ExceptionCode.

Reply:
What is the difference between these two cases? The error messages and error code names do not provide good context.